DfE.FindInformationAcademiesTrusts.Data.AcademiesDb/DataSourceProvider.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Microsoft.EntityFrameworkCore;
+
+namespace DfE.FindInformationAcademiesTrusts.Data.AcademiesDb;
+
+public class DataSourceProvider : IDataSourceProvider
+{
+    private readonly IAcademiesDbContext _academiesDbContext;
+
+    [ExcludeFromCodeCoverage]
+    public DataSourceProvider(AcademiesDbContext academiesDbContext) : this(
+        (IAcademiesDbContext)academiesDbContext)
+    {
+    }
+
+    public DataSourceProvider(IAcademiesDbContext academiesDbContext)
+    {
+        _academiesDbContext = academiesDbContext;
+    }
+
+    public async Task<DataSource?> GetGiasUpdated()
+    {
+        var lastEntry = await _academiesDbContext.ApplicationEvents
+            .Where(e => e.Source != null
+                        && e.Source.Contains("adf-t1") && e.Source.Contains("-sips-dataflow")
+                        && e.Message == "Finished"
+                        && e.EventType != 'E'
+                        && e.Description == "GIAS_Daily").MaxAsync(e => e.DateTime);
+        if (lastEntry is null) return null;
+        return new DataSource("Get information about schools", lastEntry.Value, "Daily");
+    }
+
+    public async Task<DataSource?> GetMstrUpdated()
+    {
+        var lastEntry = await _academiesDbContext.ApplicationEvents
+            .Where(e => e.Source != null
+                        && e.Source.Contains("adf-t1") && e.Source.Contains("-sips-dataflow")
+                        && e.Message == "Finished"
+                        && e.EventType != 'E'
+                        && e.Description == "MSTR_Daily").MaxAsync(e => e.DateTime);
+        if (lastEntry is null) return null;
+        return new DataSource("Get information about schools", lastEntry.Value, "Daily");
```

Comment:
At this level in the backend it feels wrong to call mstr gias - in the future people may use parts of mstr which are not from gias. Also the date/time returned from this method will clash with that from `GetGiasUpdated`.
I know that for our purposes right now they are the same thing but if we make the decision to call them the same thing clear in the UI code near the datasourcelist rather than up here then it will be clearer to future maintainers